python/ray/serve/_private/deployment_state.py

```diff
@@ -1389,6 +1390,268 @@ def __repr__(self):
         return repr(self._replicas)
 
 
+class DeploymentRankManager:
+    """Manages replica ranks for a deployment.
+    This class handles rank assignment, release, consistency checking, and reassignment.
+    It maintains the rank system invariants and provides a clean interface for rank operations.
+    """
+
+    def __init__(self, _fail_on_error: Optional[bool] = None):
+        # Maps replica_id to assigned rank
+        self._replica_ranks: Dict[str, int] = {}
+        # Set of available ranks (initially empty, grows as target replicas change)
+        self._available_ranks: Set[int] = set()
```

Comment:
same question as the other pr: why not just use `available_ranks` and get rid of `next_rank`

separately: if we do stick with using both this and `next_rank`, can we rename this to something other than `available_ranks`, which to me implies it's a comprehensive of all available ranks? something like `released_ranks` or something else that conveys it tracks gaps

Reply:
If I keep available_ranks, then I either have to prefill it with large number of ranks upfront or expand it with autoscaling which seems like too much work.

I will rename it to `released_ranks `